pkg/repair/repair.go

```diff
@@ -0,0 +1,220 @@
+
+package repair
+
+import (
+	"strings"
+
+	"istio.io/pkg/log"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	client "k8s.io/client-go/kubernetes"
+)
+
+type Options struct {
+	PodLabelKey   string `json:"pod_label_key"`
+	PodLabelValue string `json:"pod_label_value"`
+}
+
+type Filters struct {
+	NodeName                        string               `json:"node_name"`
+	SidecarAnnotation               string               `json:"sidecar_annotation"`
+	InitContainerName               string               `json:"init_container_name"`
+	InitContainerTerminationMessage string               `json:"init_container_termination_message"`
+	InitContainerExitCode           int                  `json:"init_container_exit_code"`
+	FieldSelectors                  *KeyValueSelectorSet `json:"field_selectors"`
+	LabelSelectors                  *KeyValueSelectorSet `json:"label_selectors"`
+}
+
+// A safe getter for the FieldSelectors field of a Filters struct.
+// If the FieldSelectors struct is nil, it will create an empty one.
+func (f Filters) GetFieldSelectors() *KeyValueSelectorSet {
+	if f.FieldSelectors == nil {
+		f.FieldSelectors = &KeyValueSelectorSet{}
+	}
+	return f.FieldSelectors
+}
+
+// A safe getter for the LabelSelectors field of a Filters struct.
+// If the LabelSelectors struct is nil, it will create an empty one.
+func (f Filters) GetLabelSelectors() *KeyValueSelectorSet {
+	if f.LabelSelectors == nil {
+		f.LabelSelectors = &KeyValueSelectorSet{}
+	}
+	return f.LabelSelectors
+}
+
+// A struct wrapping a slice of field selector strings, used for convenience in
+// constructing a list of field selectors
+type KeyValueSelectorSet struct {
+	KeyValueSelectors []string
+}
+
+// Adds one or more selectors in format key=value to a KeyValueSelectorSet
+func (a *KeyValueSelectorSet) AddSelectors(selectors ...string) {
+	for _, selector := range selectors {
+		if selector != "" {
+			a.KeyValueSelectors = append(a.KeyValueSelectors, selector)
+		}
+	}
+}
+
+// Returns a stringified KeyValueSelectorSet
+func (a *KeyValueSelectorSet) String() string {
+	return strings.Join(a.KeyValueSelectors, ",")
+}
+
+// The pod reconciler struct. Contains state used to reconcile broken pods.
+type BrokenPodReconciler struct {
+	client  client.Interface
+	Filters *Filters
+	Options *Options
+}
+
+// Constructs a new BrokenPodReconciler struct.
+func NewBrokenPodReconciler(client client.Interface, filters *Filters, options *Options) (bpr BrokenPodReconciler) {
+	bpr = BrokenPodReconciler{
+		client:  client,
+		Filters: filters,
+		Options: options,
+	}
+	return
+}
+
+// Label all pods detected as broken by ListPods with a customizable label
+func (bpr BrokenPodReconciler) LabelBrokenPods() (err error) {
+	// Get a list of all broken pods
+	podList, err := bpr.ListBrokenPods()
+	if err != nil {
+		return
+	}
+
+	for _, pod := range podList.Items {
+		labels := pod.GetLabels()
+		if labels == nil {
+			labels = map[string]string{}
```

Comment:
nit: labels init can be moved to "else - label not found section"

Reply:
Fixed